Add parallel mode, Path Aliases, and relative paths

AlejandroFrias opened the pull request:
This PR ports coverage's parallel mode and path aliasing. It also adds relative path support.

## Parallel Mode and [paths] aliasing:
http://coverage.readthedocs.io/en/coverage-4.2/cmd.html?highlight=parallel#cmd-combining
As seen in the docs above, coverage has few nice features for running coverage on multiple machines in parallel. Parallel mode gives unique names to the output coverage file. the `[paths]` config option allows combining coverage files from different machines, aliasing the roots to a common root. This PR ports that behavior to Smother.

## Relative Paths
Historically coverage only grabs absolute paths, even when combining data files. When combing files on a different machine than the file will be used on, it makes sense to have relative paths so that no extra steps need to be taken to create nice html coverage files. So I added that as a possible flag. Both on combine and initial run, the paths are converted so that they're relative to the current directory that the command is run from.

# TODO
Could use more tests. I've only added one basic test that captures one common use case of combining files and using relative paths. That isn't robust enough. Not sure how to test the unique naming when run form different machines, yet.

AlejandroFrias commented:
@ChrisBeaumont, any ideas of how to test the parallel mode that names files uniquely based on process id and host name? Coverage does it, but it's a little more complicated than we need, I think. Maybe I could patch the functions? (`socket.gethostname, `os.getpid`, `random.randint`)

ChrisBeaumont commented on Makefile:
```diff
@@ -2,7 +2,10 @@
 
-test:
+clean:
+	find . -name "*.pyc" -exec rm {} +
```
what do you think about using `git clean -fxd` instead?

AlejandroFrias replied:
Not great for development. It'll remove any untracked files I'm working on and testing with, but haven't committed yet. That's no fun. I can (and probably should) build this out a bit more to remove all build related files (so `__pycache__/`, `build/`, `dist/`,`htmlcov/`, `smother.egg-info/`, `.cache/`). Did I miss any there?

AlejandroFrias replied:
You also wouldn't want to delete their swp files or any other temporary files created by text editors.

ChrisBeaumont replied:
SGTM

ChrisBeaumont commented on smother/control.py:
```diff
@@ -81,6 +86,13 @@ def write(self, file_or_path, append=False, timeout=10):
         multiprocess test environment.
         """
         if isinstance(file_or_path, six.string_types):
+            if self.coverage and self.coverage.config.parallel:
```
my first instinct is that `write()` should be "dumb" and treat file_or_path as litearally as possible. Can we move this logic up to a helper function invoked by the caller of `write`? Or do you think it makes more sense here?

AlejandroFrias replied:
I was imitating how coverage does it. They wait until the last possible moment (which is during write) to determine the output file name. It made sense to do it that way in the coverage source so that it wouldn't interfere elsewhere, but smother isn't as complicated, so it is possible to determine the filename later. I like it this way still, though, because write is called from a lot of places and it makes sense to me that the output filename should be automatically modified if the parallel option is being used.

ChrisBeaumont replied:
Ok that SGTM -- let's just update the docstring to note when and how file_or_path is renamed

AlejandroFrias replied:
In append mode does it expect `file_or_path` to already exist? That might be slightly problematic if so.

ChrisBeaumont commented on smother/control.py:
```diff
@@ -119,9 +131,25 @@ def load(cls, file_or_path):
+            path = aliases.map(path)
+
+        if self.relative_paths:
+            set_relative_directory()
```
what side effect does this function have?

AlejandroFrias replied:
https://bitbucket.org/ned/coveragepy/src/7c3dddfe83893c51035c0be4f35b58d0379858ac/coverage/files.py?at=default&fileviewer=file-view-default

Sets some icky globals that are used to create the relative path int he next line. Not the best, but already well tested since it's from coverage.

ChrisBeaumont commented on smother/control.py:
```diff
@@ -41,7 +46,7 @@ def start(self):
     def save_context(self, label):
         self.data[label] = {
-            key: sorted(map(int, val.keys()))
+            self._normalize_path(key): sorted(map(int, val.keys()))
```
so if I'm following so far we are adding extra logic at coverage write time, so that the data files have normalized (potentially relative) paths that can then be interpreted literally at read time. Is that right? Have you considered switching that around, so that writing the reports is "simple" and the extra normalization logic happens when the data file is read?

I'm not sure which would be better. One nice thing about deferring the logic is that you can more easily correct mistakes / refine your path aliases once you've written a bunch of reports -- if path aliases rely on collection-time behavior then any mistakes require re-running the test suite.

OTOH, files are read more often than they are written, so maybe that argues for keeping the logic at write-time

ChrisBeaumont replied:
One other though: if the path alias logic were applied at read time instead of write time, I think we could eliminate the `--relative-paths` option -- report data would always be an absolute path, and the `.coveragerc` file could specify how to normalize those paths (potentially to relative paths if desired).

AlejandroFrias replied:
When combining smother files, normalizing the paths at write time is necessary and is what I do. But you're correct that at write time when generating smother files it is not necessary to do anything to path. I just thought that it was nicer to have the option to create a smother file with cleaned paths right off the bat when you want. But the aliasing probably shouldn't happen at this stage. That should only be used for combining files.

AlejandroFrias replied:
>One other thought: if the path alias logic were applied at read time instead of write time, I think we could eliminate the `--relative-paths` option -- report data would always be an absolute path, and the .coveragerc file could specify how to normalize those paths (potentially to relative paths if desired).

I have to add the `--relative-paths` option at some point no matter what. The way coverage's `PathAliases` class works, it always converts to abosolute paths, regardless of what the aliases are given. So if want to be able to produce a smother file, either by combining several or just regularly, that has relative paths, I have to call the function. It doesn't make sense to do it at read time, because it would be very expensive to the lookup.

ChrisBeaumont replied:
Ok this thread is my only remaining concern. I'd prefer that we find a way to do away with the `--reative-paths` flag since it complicates the user-facing interface and seems too similar to what the `paths` config section already does. Let's refine the terminology and goals here. There are 3 phases of the .smother file lifecycle: *collection* time, *combine* time, and *report* time. I think they should work like this

* At *collect* time, we always do what coverage does (we dump absolute paths to a file).
* At *combine time*, we've already gathered several .smother files onto a new machine. We can assume that this machine will be used during report time, and the files referenced in the smother report are available somewhere (we're often but not always in the root directory of the files in a .smother file). The user is required to configure their coveragerc file so that the path section normalizes files into correct paths on the current machine. If we use PathAliases the combined report will [very likely include](https://bitbucket.org/ned/coveragepy/src/7c3dddfe83893c51035c0be4f35b58d0379858ac/coverage/files.py?at=default&fileviewer=file-view-default#files.py-352) absolute paths, but that's ok. Alternatively we could implement our own PathAliases equivalent that doesn't convert to canonical paths, and instead treats the aliases in the coveragerc file more literally. 
* At *report time* we don't care whether the input data are absolute or relative paths, we just require that `os.path.exists(path)`. We don't apply any normalization, so we don't pay any speed penalties. 

This flow allows us to eliminate the `--relative-paths` flag, and the only thing we lose is the ability to force smother files that have relative paths. Relative paths are kind of nice if you're manually looking at the file, but not required for portability since you can always achieve the same thing with a `.coveragerc` and `smother combine`. I don't think that's a strong enough reason to add a new flag to smother.

AlejandroFrias replied:
> Relative paths are kind of nice if you're manually looking at the file, but not required for portability since you can always achieve the same thing with a .coveragerc and smother combine. I don't think that's a strong enough reason to add a new flag to smother.

This is only true if we make our own PathAliases that doesn't use the canonical path. Otherewise there is no way (without editing the file ourselves) to get relative paths. We could do that, but that might mess with people's assumptions that it has the same behavior as coverage. What if we have a keep the `--relative-paths` option only for `combine` as a nice bonus feature?

ChrisBeaumont replied:
> there is no way (without editing the file ourselves) to get relative paths

You are right -- I just meant you can achieve the "same thing" in that you can create a correct smother file where the paths resolve on the local filesystem, without the flag. The only difference is that the paths themselves are absolute.

I'm conceptually on board with the rest of this PR. If you still feel strongly about `--relative-paths`, can we move that to a dedicated PR and focus this one on implementing path aliasing without the flag?

ChrisBeaumont commented:
> any ideas of how to test the parallel mode that names files uniquely based on process id and host name?

I would probably extract the logic into a helper function with a signature like `def _get_smother_filename(base_name, parallel_mode)`, and call that within the `write()` method. You can then test that function without actually writing any files.

ChrisBeaumont commented on smother/control.py:
```diff
@@ -119,9 +131,25 @@ def load(cls, file_or_path):
+    def _normalize_path(self, path):
+        aliases = None
+        if self.coverage and self.coverage.config.paths:
+            aliases = PathAliases()
```
We should probably only build a PathAliases() object once, when we write the coverage report to disk. It involves compiling a bunch of regexes.

ChrisBeaumont replied:
Or better yet, we refactor this into a `_normalize` helper function with no inputs, that normalizes the entire in-memory smother report. Then we only invoke this function when we call `smother combine`, and not during first collection.

AlejandroFrias commented:
@ChrisBeaumont, How's this look?

I removed the `--relative-paths` flag and added a utility command for converting a smother file to relative paths.

ChrisBeaumont commented on smother/cli.py:
```diff
@@ -17,12 +17,22 @@
     is_flag=True,
 )
+@click.option(
+    '--coveragerc',
```
let's call this 'rcfile' to [mirror coverage](https://coverage.readthedocs.io/en/coverage-4.2/config.html?highlight=coveragerc#configuration-files)

ChrisBeaumont commented on smother/control.py:
```diff
@@ -205,3 +240,14 @@ def iter_records(self, semantic=False, sort=True):
+    def _normalize_combine_path(self, path):
+        aliases = None
+        if self.coverage and self.coverage.config.paths:
+            aliases = PathAliases()
```
Let's only build PathAliases once within `__ior__`, instead of every time we call this function

ChrisBeaumont commented on smother/tests/test_controller.py:
```diff
@@ -72,3 +74,25 @@ def test_iter_records_semantic():
+
+
+def test_convert_to_relative_paths():
+    pass
```
todo?

ChrisBeaumont commented:
Nice, :+1: on the approach now. I left a few small notes, but LGTM otherwise

AlejandroFrias commented:
@ChrisBeaumont, cool. I think that last couple commits should cover your comments. I created the PathAliases object in `__init__` so it will only be created once and finished the tests for the relative paths. Let's merge this if it still LGTY.

ChrisBeaumont commented on smother/control.py:
```diff
@@ -142,21 +153,21 @@ def load(cls, file_or_path):
     def convert_to_relative_paths(cls, smother_obj):
-        data = {}
+        data = defaultdict(lambda: dict())
         set_relative_directory()
```
Can we git rid of set_relative_directory() now, or do we still need it?

ChrisBeaumont replied:
nvm, I think we do

ChrisBeaumont commented:
LGTM! This is great, thank
